cpp/src/arrow/stl_iterator.h

```diff
@@ -128,6 +131,215 @@ class ArrayIterator {
+      index_ += n;
+      iterators_list_[current_chunk_index_] += n;
+      return *this;
+    } else {
+      iterators_list_[current_chunk_index_] -=
+          iterators_list_[current_chunk_index_].index();
+      index_ += n;
+      InitializeComponents(*this, true);
+      return *this;
+    }
+  }
+  ChunkedArrayIterator& operator-=(difference_type n) {
+    if (n <= iterators_list_[current_chunk_index_].index()) {
+      index_ -= n;
+      iterators_list_[current_chunk_index_] -= n;
+      return *this;
+    } else {
+      iterators_list_[current_chunk_index_] -=
+          iterators_list_[current_chunk_index_].index();
+      index_ -= n;
+      InitializeComponents(*this, true);
+      return *this;
+    }
+  }
+
+  // Comparisons
+  bool operator==(const ChunkedArrayIterator& other) const {
+    return index_ == other.index_;
+  }
+  bool operator!=(const ChunkedArrayIterator& other) const {
+    return index_ != other.index_;
+  }
+  bool operator<(const ChunkedArrayIterator& other) const {
+    return index_ < other.index_;
+  }
+  bool operator>(const ChunkedArrayIterator& other) const {
+    return index_ > other.index_;
+  }
+  bool operator<=(const ChunkedArrayIterator& other) const {
+    return index_ <= other.index_;
+  }
+  bool operator>=(const ChunkedArrayIterator& other) const {
+    return index_ >= other.index_;
+  }
+
+ private:
+  int64_t GetChunkIndex(int64_t index) const {
+    return static_cast<int64_t>(
+        std::upper_bound(chunks_lengths_.begin(), chunks_lengths_.end(), index) -
+        chunks_lengths_.begin());
+  }
+
+  void InitializeComponents(ChunkedArrayIterator<ArrayType>& chunked_array_iterator,
+                            bool update = false) {
+    if (!update) {
+      int64_t chunk_index = 0;
+      for (const auto& array : chunked_array_iterator.chunked_array_->chunks()) {
+        chunked_array_iterator.iterators_list_.emplace_back(
+            *arrow::internal::checked_pointer_cast<ArrayType>(array));
+        auto chunk_length = static_cast<int64_t>(array->length());
+        if (chunk_index) {
+          chunked_array_iterator.chunks_lengths_.push_back(
+              chunk_length + chunked_array_iterator.chunks_lengths_[chunk_index - 1]);
+        } else {
+          chunked_array_iterator.chunks_lengths_.push_back(chunk_length);
+        }
+        chunk_index++;
+      }
+    }
+
+    chunked_array_iterator.current_chunk_index_ =
+        GetChunkIndex(chunked_array_iterator.index_);
+    auto& current_iterator =
+        chunked_array_iterator
+            .iterators_list_[chunked_array_iterator.current_chunk_index_];
+    current_iterator -= current_iterator.index();
+    if (chunked_array_iterator.current_chunk_index_)
+      current_iterator +=
+          chunked_array_iterator.index_ -
+          chunked_array_iterator
+              .chunks_lengths_[chunked_array_iterator.current_chunk_index_ - 1];
+    else
+      current_iterator += chunked_array_iterator.index_;
+  }
```

Comment:
Not sure if this is enforced in Arrow but I suggest to always use curly-braces for `if-else` blocks.
Nit: This can be simplified to
```c++
auto& current_iterator = chunked_array_iterator.iterators_list_[chunked_array_iterator.current_chunk_index_] - current_iterator.index() + chunked_array_iterator.index_;
if (chunked_array_iterator.current_chunk_index_) {
  current_iterator += chunked_array_iterator.chunks_lengths_[chunked_array_iterator.current_chunk_index_ - 1];
}
```

Reply:
I agree with this suggestion for the record. We don't enforce it but we try to maintain it.